Panic in Datadog backend

```
panic: reflect: slice index out of range [recovered]
#011panic: reflect: slice index out of range
goroutine 349 [running]:
panic(0x9ac080, 0xc820407aa0)
#011/usr/local/Cellar/go/1.6.2/libexec/src/runtime/panic.go:481 +0x3e6
encoding/json.(*encodeState).marshal.func1(0xc820205cb8)
#011/usr/local/Cellar/go/1.6.2/libexec/src/encoding/json/encode.go:269 +0x11f
panic(0x9ac080, 0xc820407a90)
#011/usr/local/Cellar/go/1.6.2/libexec/src/runtime/panic.go:443 +0x4e9
reflect.Value.Index(0x9927a0, 0xc82023f9e0, 0x197, 0x175, 0x0, 0x0, 0x0)
#011/usr/local/Cellar/go/1.6.2/libexec/src/reflect/value.go:854 +0x151
encoding/json.(*arrayEncoder).encode(0xc82001e1e8, 0xc820166000, 0x9927a0, 0xc82023f9e0, 0x197, 0x0)
#011/usr/local/Cellar/go/1.6.2/libexec/src/encoding/json/encode.go:690 +0xd2
encoding/json.(*arrayEncoder).(encoding/json.encode)-fm(0xc820166000, 0x9927a0, 0xc82023f9e0, 0x197, 0x0)
#011/usr/local/Cellar/go/1.6.2/libexec/src/encoding/json/encode.go:697 +0x51
encoding/json.(*sliceEncoder).encode(0xc82001e1f0, 0xc820166000, 0x9927a0, 0xc82023f9e0, 0x197, 0x0)
#011/usr/local/Cellar/go/1.6.2/libexec/src/encoding/json/encode.go:667 +0xb0
encoding/json.(*sliceEncoder).(encoding/json.encode)-fm(0xc820166000, 0x9927a0, 0xc82023f9e0, 0x197, 0x0)
#011/usr/local/Cellar/go/1.6.2/libexec/src/encoding/json/encode.go:676 +0x51
encoding/json.(*structEncoder).encode(0xc82016a330, 0xc820166000, 0xb2a980, 0xc82023f9e0, 0x199, 0x0)
#011/usr/local/Cellar/go/1.6.2/libexec/src/encoding/json/encode.go:587 +0x2c4
encoding/json.(*structEncoder).(encoding/json.encode)-fm(0xc820166000, 0xb2a980, 0xc82023f9e0, 0x199, 0xc82023f900)
#011/usr/local/Cellar/go/1.6.2/libexec/src/encoding/json/encode.go:601 +0x51
encoding/json.(*ptrEncoder).encode(0xc82001e1f8, 0xc820166000, 0xacaca0, 0xc82023f9e0, 0x16, 0x0)
#011/usr/local/Cellar/go/1.6.2/libexec/src/encoding/json/encode.go:709 +0xea
encoding/json.(*ptrEncoder).(encoding/json.encode)-fm(0xc820166000, 0xacaca0, 0xc82023f9e0, 0x16, 0xc82023f900)
#011/usr/local/Cellar/go/1.6.2/libexec/src/encoding/json/encode.go:714 +0x51
encoding/json.(*encodeState).reflectValue(0xc820166000, 0xacaca0, 0xc82023f9e0, 0x16)
#011/usr/local/Cellar/go/1.6.2/libexec/src/encoding/json/encode.go:301 +0x6b
encoding/json.(*encodeState).marshal(0xc820166000, 0xacaca0, 0xc82023f9e0, 0x0, 0x0)
#011/usr/local/Cellar/go/1.6.2/libexec/src/encoding/json/encode.go:274 +0xa9
encoding/json.Marshal(0xacaca0, 0xc82023f9e0, 0x0, 0x0, 0x0, 0x0, 0x0)
#011/usr/local/Cellar/go/1.6.2/libexec/src/encoding/json/encode.go:139 +0x84
github.com/atlassian/gostatsd/backend/backends/datadog.(*client).post(0xc82000b180, 0xc1bc50, 0xe, 0xc15088, 0x7, 0xacaca0, 0xc82023f9e0, 0x0, 0x0)
#011/Users/mmazurskiy/gopath/src/github.com/atlassian/gostatsd/backend/backends/datadog/datadog.go:229 +0x52
github.com/atlassian/gostatsd/backend/backends/datadog.(*client).postMetrics(0xc82000b180, 0xc82023f9e0, 0x0, 0x0)
#011/Users/mmazurskiy/gopath/src/github.com/atlassian/gostatsd/backend/backends/datadog/datadog.go:200 +0x7b
github.com/atlassian/gostatsd/backend/backends/datadog.(*client).SendMetricsAsync.func1.1(0xc82000b180, 0xc82023f9e0, 0x7f249cdb4ee0, 0xc82003ab40, 0xc820458120)
#011/Users/mmazurskiy/gopath/src/github.com/atlassian/gostatsd/backend/backends/datadog/datadog.go:120 +0x3d
created by github.com/atlassian/gostatsd/backend/backends/datadog.(*client).SendMetricsAsync.func1
#011/Users/mmazurskiy/gopath/src/github.com/atlassian/gostatsd/backend/backends/datadog/datadog.go:125 +0x6f
Jun 22 00:37:05 ip-10-116-8-214 systemd[1]: statsd.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
```

--- a/Makefile
+++ b/Makefile
@@ -9,6 +9,7 @@ BINARY_NAME := gostatsd
 IMAGE_NAME := atlassianlabs/$(BINARY_NAME)
 ARCH ?= darwin
 METALINTER_CONCURRENCY ?= 4
+GOVERSION = 1.6.2
 
 setup: setup-ci
 	go get -v -u github.com/githubnemo/CompileDaemon
@@ -83,20 +84,38 @@ watch:
 git-hook:
 	cp dev/push-hook.sh .git/hooks/pre-push
 
+# Compile a static binary. Cannot be used with -race
 cross:
-	CGO_ENABLED=0 GOOS=linux go build -o build/bin/linux/$(BINARY_NAME) $(GOBUILD_VERSION_ARGS) -a -installsuffix cgo  github.com/atlassian/$(BINARY_NAME)
+	CGO_ENABLED=0 GOOS=linux go build -o build/bin/linux/$(BINARY_NAME) $(GOBUILD_VERSION_ARGS) -a -installsuffix cgo github.com/atlassian/gostatsd
 
 docker: cross
-	cd build && docker build --pull -t $(IMAGE_NAME):$(GIT_HASH) .
+	docker build --pull -t $(IMAGE_NAME):$(GIT_HASH) build
+
+# Compile a binary with -race.
+docker-race:
+	docker run \
+		--rm \
+		-v "$(GOPATH)":/opt/gopath \
+		-w /opt/gopath/src/github.com/atlassian/gostatsd \
+		-e GOPATH=/opt/gopath \
+		golang:$(GOVERSION) \
+		go build -race -o build/bin/linux/$(BINARY_NAME) $(GOBUILD_VERSION_ARGS) -a -installsuffix cgo github.com/atlassian/gostatsd
+	docker build --pull -t $(IMAGE_NAME):$(GIT_HASH)-race -f build/Dockerfile-glibc build
 
 release-hash: check test docker
 	docker push $(IMAGE_NAME):$(GIT_HASH)
 
-release: check test docker release-hash
+release-normal: release-hash
 	docker tag $(IMAGE_NAME):$(GIT_HASH) $(IMAGE_NAME):latest
 	docker push $(IMAGE_NAME):latest
 	docker tag $(IMAGE_NAME):$(GIT_HASH) $(IMAGE_NAME):$(REPO_VERSION)
 	docker push $(IMAGE_NAME):$(REPO_VERSION)
+
+release-race: docker-race
+	docker tag $(IMAGE_NAME):$(GIT_HASH)-race $(IMAGE_NAME):$(REPO_VERSION)-race
+	docker push $(IMAGE_NAME):$(REPO_VERSION)-race
+
+release: release-normal release-race
 
 run: build
 	./build/bin/$(ARCH)/$(BINARY_NAME) --backends=stdout --verbose --flush-interval=1s

--- a/backend/backends/datadog/datadog.go
+++ b/backend/backends/datadog/datadog.go
@@ -113,7 +113,7 @@ func (d *client) SendMetricsAsync(ctx context.Context, metrics *types.MetricMap,
 		cb(nil)
 		return
 	}
-	var counter uint16
+	counter := 0
 	results := make(chan error)
 	d.processMetrics(metrics, func(ts *timeSeries) {
 		go func() {
@@ -128,7 +128,7 @@ func (d *client) SendMetricsAsync(ctx context.Context, metrics *types.MetricMap,
 	go func() {
 		errs := make([]error, 0, counter)
 	loop:
-		for c := counter; c > 0; c-- {
+		for c := 0; c < counter; c++ {
 			select {
 			case <-ctx.Done():
 				errs = append(errs, ctx.Err())
@@ -151,7 +151,7 @@ func (d *client) processMetrics(metrics *types.MetricMap, cb func(*timeSeries)) 
 	metrics.Counters.Each(func(key, tagsKey string, counter types.Counter) {
 		ts.addMetric(key, tagsKey, rate, counter.PerSecond, counter.Flush)
 		ts.addMetric(fmt.Sprintf("%s.count", key), tagsKey, gauge, float64(counter.Value), counter.Flush)
-		d.maybeFlush(ts, cb)
+		d.maybeFlush(&ts, cb)
 	})
 
 	metrics.Timers.Each(func(key, tagsKey string, timer types.Timer) {
@@ -167,17 +167,17 @@ func (d *client) processMetrics(metrics *types.MetricMap, cb func(*timeSeries)) 
 		for _, pct := range timer.Percentiles {
 			ts.addMetric(fmt.Sprintf("%s.%s", key, pct.Str), tagsKey, gauge, pct.Float, timer.Flush)
 		}
-		d.maybeFlush(ts, cb)
+		d.maybeFlush(&ts, cb)
 	})
 
 	metrics.Gauges.Each(func(key, tagsKey string, g types.Gauge) {
 		ts.addMetric(key, tagsKey, gauge, g.Value, g.Flush)
-		d.maybeFlush(ts, cb)
+		d.maybeFlush(&ts, cb)
 	})
 
 	metrics.Sets.Each(func(key, tagsKey string, set types.Set) {
 		ts.addMetric(key, tagsKey, gauge, float64(len(set.Values)), set.Flush)
-		d.maybeFlush(ts, cb)
+		d.maybeFlush(&ts, cb)
 	})
 
 	if len(ts.Series) > 0 {
@@ -185,13 +185,15 @@ func (d *client) processMetrics(metrics *types.MetricMap, cb func(*timeSeries)) 
 	}
 }
 
-func (d *client) maybeFlush(ts *timeSeries, cb func(*timeSeries)) {
-	if uint(len(ts.Series))+20 >= d.metricsPerBatch { // flush before it reaches max size and grows the slice
-		cb(ts)
-		*ts = timeSeries{
+func (d *client) maybeFlush(ts **timeSeries, cb func(*timeSeries)) {
+	t := *ts
+	if uint(len(t.Series))+20 >= d.metricsPerBatch { // flush before it reaches max size and grows the slice
+		cb(t)
+		// We want external ts pointer to point to the new object
+		*ts = &timeSeries{
 			Series:    make([]metric, 0, d.metricsPerBatch),
-			Timestamp: ts.Timestamp,
-			Hostname:  ts.Hostname,
+			Timestamp: t.Timestamp,
+			Hostname:  t.Hostname,
 		}
 	}
 }

--- a/backend/backends/datadog/datadog_test.go
+++ b/backend/backends/datadog/datadog_test.go
@@ -4,6 +4,7 @@ import (
 	"io/ioutil"
 	"net/http"
 	"net/http/httptest"
+	"sync/atomic"
 	"testing"
 	"time"
 
@@ -13,11 +14,11 @@ import (
 )
 
 func TestRetries(t *testing.T) {
-	var requestNum int
+	var requestNum uint32
 	mux := http.NewServeMux()
 	mux.HandleFunc("/api/v1/series", func(w http.ResponseWriter, r *http.Request) {
 		defer r.Body.Close()
-		requestNum++
+		atomic.AddUint32(&requestNum, 1)
 		data, err := ioutil.ReadAll(r.Body)
 		if err != nil {
 			t.Error(err)
@@ -31,7 +32,7 @@ func TestRetries(t *testing.T) {
 		}
 		if requestNum == 1 {
 			// Return error on first request to trigger a retry
-			w.WriteHeader(400)
+			w.WriteHeader(http.StatusBadRequest)
 		}
 	})
 	ts := httptest.NewServer(mux)
@@ -56,12 +57,57 @@ func TestRetries(t *testing.T) {
 	}
 }
 
+func TestSendMetrics(t *testing.T) {
+	var requestNum uint32
+	mux := http.NewServeMux()
+	mux.HandleFunc("/api/v1/series", func(w http.ResponseWriter, r *http.Request) {
+		defer r.Body.Close()
+		atomic.AddUint32(&requestNum, 1)
+		data, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			t.Error(err)
+			return
+		}
+		if len(data) == 0 {
+			t.Errorf("empty body")
+			w.WriteHeader(http.StatusBadRequest)
+		}
+		if r.ContentLength != int64(len(data)) {
+			t.Errorf("unexpected body length: %d. Content-Length is %d", len(data), r.ContentLength)
+			w.WriteHeader(http.StatusBadRequest)
+		}
+	})
+	ts := httptest.NewServer(mux)
+	defer ts.Close()
+
+	client, err := NewClient(ts.URL, "apiKey123", 1, 1*time.Second, 2*time.Second)
+	if err != nil {
+		t.Fatal(err)
+	}
+	res := make(chan []error, 1)
+	client.SendMetricsAsync(context.Background(), metrics(), func(errs []error) {
+		res <- errs
+	})
+	errs := <-res
+	for _, err := range errs {
+		if err != nil {
+			t.Error(err)
+		}
+	}
+	if requestNum != 2 {
+		t.Errorf("unexpected number of requests: %d", requestNum)
+	}
+}
+
 func metrics() *types.MetricMap {
 	return &types.MetricMap{
-		NumStats: 1,
+		NumStats: 2,
 		Counters: types.Counters{
 			"stat1": map[string]types.Counter{
 				"tag1": types.NewCounter(time.Now(), 1*time.Second, 5),
+			},
+			"stat2": map[string]types.Counter{
+				"tag2": types.NewCounter(time.Now(), 2*time.Second, 50),
 			},
 		},
 	}

--- a/build/Dockerfile-glibc
+++ b/build/Dockerfile-glibc
@@ -1,0 +1,10 @@
+FROM ubuntu:16.04
+
+RUN export DEBIAN_FRONTEND=noninteractive && \
+    apt-get update && \
+    apt-get install ca-certificates -y && \
+    rm -rf /var/lib/apt/lists/*
+
+ADD bin/linux/gostatsd /bin/gostatsd
+
+ENTRYPOINT ["gostatsd"]

--- a/tester/fakesocket/fake.go
+++ b/tester/fakesocket/fake.go
@@ -15,7 +15,7 @@ var FakeMetric = []byte("foo.bar.baz:2|c")
 
 // FakeAddr is a fake net.Addr
 var FakeAddr = &net.UDPAddr{
-	IP: net.IPv4(127, 0, 0, 1),
+	IP:   net.IPv4(127, 0, 0, 1),
 	Port: 8181,
 }
 

